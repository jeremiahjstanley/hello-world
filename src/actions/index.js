export const setLocation = (location) => ({
  type: 'SET_LOCATION',
  location
});

export const setDataSet = (dataSet) => ({
  type: 'SET_DATASET',
  dataSet
});

export const clearDataSet = () => ({
  type: 'CLEAR_DATASET',
});

export const setDataBase = (dataBase) => ({
  type: 'SET_DATABASE',
  dataBase
});

export const fetchLocationDataSuccess = (locationData) => ({
  type: 'FETCH_LOCATION_DATA_SUCCESS',
  locationData
});

export const hasErrored = (status) => ({
  type: 'HAS_ERRORED',
  hasErrored: status
});

export const madeSearch = (status) => ({
  type: 'MADE_SEARCH',
  madeSearch: status
});


