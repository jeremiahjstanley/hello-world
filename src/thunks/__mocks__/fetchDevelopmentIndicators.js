export const fetchDevelopmentIndicators = jest.fn().mockImplementation(() => {
  return {};
});