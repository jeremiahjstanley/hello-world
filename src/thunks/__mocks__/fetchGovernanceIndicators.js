export const fetchGovernanceIndicators = jest.fn().mockImplementation(() => {
  return {};
});