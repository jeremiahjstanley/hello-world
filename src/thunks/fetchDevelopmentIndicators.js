import { fetchLocationDataSuccess } from '../actions';
import { dataCleaner } from '../helper/dataCleaner';
import { hasErrored } from '../actions';
import { key } from '../helper/apiKey';

export const fetchDevelopmentIndicators = (locations, dataSet, dataBase) => {

    return async (dispatch) => {
        const unresolvedDevelopmentIndicators = locations.map(async location => {
            const isoAlpha3 = location.alpha_3
            try {
              const url = `https://www.quandl.com/api/v3/datasets/${dataBase}/${isoAlpha3}_${dataSet}.json?api_key=${key}`;
              const response = await fetch(url);
              if (!response.ok) {
                throw Error(response.statusText);
              }
              return await response.json();
            } catch (error) {
              dispatch(hasErrored(true));
            }
        });

        const developmentIndicators = await Promise.all(unresolvedDevelopmentIndicators);
        const cleanDevelopmentIndicators = developmentIndicators.map(indicator => {
            return dataCleaner(indicator);
        });
        
        await dispatch(fetchLocationDataSuccess(cleanDevelopmentIndicators))   
    };
};