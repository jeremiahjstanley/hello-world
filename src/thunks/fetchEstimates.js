import { hasErrored } from '../actions';
import { key } from '../helper/apiKey';

export const fetchEstimates = (isoAlpha3, dataSet, dataBase) => {
  return async (dispatch) => {
    try {
      const url = `https://www.quandl.com/api/v3/datasets/${dataBase}/${isoAlpha3}_${dataSet}_EST.json?api_key=${key}`;
      const response = await fetch(url);
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return await response.json();
    } catch (error) {
      dispatch(hasErrored(true));
    }
  }
};