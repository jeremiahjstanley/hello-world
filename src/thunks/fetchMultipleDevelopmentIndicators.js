import { fetchAdditionalDataSuccess, hasErrored  } from '../actions';
import { dataCleaner } from '../helper/dataCleaner';
import { key } from '../helper/apiKey';

export const fetchMultipleDevelopmentIndicators = (locations, dataSets, dataBase) => {

  return async (dispatch) => {

    const unresolvedMultipleDevelopmentIndicators = dataSets.map(async dataSet => {
      const dataSetCode = dataSet.dataset_code;

      const unresolvedDevelopmentIndicators = locations.map(async location => {
        const isoAlpha3 = location.alpha_3;
        try {
          const url = `https://www.quandl.com/api/v3/datasets/${dataBase}/${isoAlpha3}_${dataSetCode}.json?api_key=${key}`;
          console.log(url)
          const response = await fetch(url);
          if (!response.ok) {
            throw Error(response.statusText);
          }
          return await response.json();
        } catch (error) {
          dispatch(hasErrored(true));
        }
      });

      const developmentIndicators = await Promise.all(unresolvedDevelopmentIndicators);
      const cleanDevelopmentIndicators = developmentIndicators.map(indicator => {
        return dataCleaner(indicator);
      });

      return cleanDevelopmentIndicators
    });

    const multipleDevelopmentIndicators = await Promise.all(unresolvedMultipleDevelopmentIndicators);
    const convertedMultipleDevelomentIndicators = multipleDevelopmentIndicators.reduce((convertedArray, indicator) => {
      convertedArray = [...convertedArray, ...indicator]
      return convertedArray
    },[]);
    await dispatch(fetchAdditionalDataSuccess(convertedMultipleDevelomentIndicators));  
  }
};
