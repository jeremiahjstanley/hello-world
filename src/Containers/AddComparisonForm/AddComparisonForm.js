import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { setLocation } from '../../actions';
import { countries } from '../../helper/countryMetrics';
import { fetchGovernanceIndicators } from '../../thunks/fetchGovernanceIndicators';
import { fetchDevelopmentIndicators } from '../../thunks/fetchDevelopmentIndicators';

export class AddComparisonForm extends Component {
  constructor() {
    super()

    this.state = {
    };

  };

  handleChange = (event) => {
    const input = event.target.value;
    const additionalLocation = countries.find(country => input === country.name);
    this.setState({additionalLocation, input});
  };

  handleSubmit = (event) => {
    event.preventDefault();
    const { location, dataSet, dataBase} = this.props;
    if (this.state.additionalLocation) {
      const locations = [...location, this.state.additionalLocation];
      this.props.selectLocation(locations);
      ((dataBase === 'WWGI') ? 
        this.props.fetchGovernanceIndicators(locations, dataSet, dataBase) : 
        this.props.fetchDevelopmentIndicators(locations, dataSet, dataBase));
      this.props.history.push('/stats');
    }
  };

  countryTypeAhead = () => {
    const suggestions = countries.filter(country => {
      const regex = new RegExp(this.state.input, 'gi');
      return country.name.match(regex) || country.alpha_3.match(regex);
    });
    return suggestions.splice(0, 4).map((country, index) => {
        return <option key={index}>{country.name}</option>
    });
  };

  render() {

    return (

      <div className='change-form-container'>
        <Link to='/stats' className='back-button'>â—€ <span>back</span></Link>
        <form onSubmit={this.handleSubmit} className='change-form'>
          <input
            type='text'
            placeholder='enter a country'
            onChange={this.handleChange}
            list="countries"
          />
          <datalist id="countries">
            { this.countryTypeAhead() }
          </datalist>
          <button 
            disabled={!this.state.additionalLocation}
            className='change-form-button'
          >Add</button>
       </form>
     </div>

    );
  };
};

export const mapStateToProps = (state) => ({
  dataBase: state.dataBase.database_code,
  dataSet: state.dataSet.dataset_code,
  location: state.location
});

export const mapDispatchToProps = (dispatch) => ({
  fetchDevelopmentIndicators: (locations, dataBase, dataSet) => dispatch(fetchDevelopmentIndicators(locations, dataBase, dataSet)),
  fetchGovernanceIndicators: (locations, dataBase, dataSet) => dispatch(fetchGovernanceIndicators(locations, dataBase, dataSet)),
  selectLocation: (location) => dispatch(setLocation(location))
});

AddComparisonForm.propTypes = {
  dataBase: PropTypes.string.isRequired,
  dataSet: PropTypes.string.isRequired,
  location: PropTypes.array.isRequired,
  fetchDevelopmentIndicators: PropTypes.func.isRequired,
  fetchGovernanceIndicators: PropTypes.func.isRequired,
  selectLocation: PropTypes.func.isRequired
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(AddComparisonForm));
