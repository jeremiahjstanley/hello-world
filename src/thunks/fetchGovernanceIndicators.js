import { fetchLocationDataSuccess } from '../actions';
import { dataCleaner } from '../helper/dataCleaner';
import { fetchEstimates } from './fetchEstimates';
import { fetchNumberOfSources } from './fetchNumberOfSources'; 
import { fetchPercentileRank } from './fetchPercentileRank';
import { fetchStandardError } from './fetchStandardError';

export const fetchGovernanceIndicators = (locations, dataSet, dataBase) => {
    return async (dispatch) => {
        const unresolvedGovernanceIndicators = locations.map(async location => {
            const alpha3 = location.alpha_3
            const estimates = await dispatch(fetchEstimates(alpha3, dataSet, dataBase));
            const numberOfSources = await dispatch(fetchNumberOfSources(alpha3, dataSet, dataBase)); 
            const percentileRank = await dispatch(fetchPercentileRank(alpha3, dataSet, dataBase));
            const standardError = await dispatch(fetchStandardError(alpha3, dataSet, dataBase));
            const cleanEstimates = dataCleaner(estimates);
            const cleanNumberOfSources = dataCleaner(numberOfSources);
            const cleanPercentileRank = dataCleaner(percentileRank);
            const cleanStandardError = dataCleaner(standardError);
            return {cleanEstimates, cleanNumberOfSources, cleanPercentileRank, cleanStandardError};
        });
        const governanceIndicators = await Promise.all(unresolvedGovernanceIndicators)
        await dispatch(fetchLocationDataSuccess(governanceIndicators))     
    };
};