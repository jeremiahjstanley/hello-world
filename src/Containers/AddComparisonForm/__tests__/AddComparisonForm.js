import React from 'react';
import { shallow } from 'enzyme';
import { AddComparisonForm, mapStateToProps, mapDispatchToProps } from '../AddComparisonForm';
import { setLocation} from '../../../actions';
import { fetchGovernanceIndicators } from '../../../thunks/fetchGovernanceIndicators';
import { fetchDevelopmentIndicators } from '../../../thunks/fetchDevelopmentIndicators';

jest.mock('../../../thunks/fetchGovernanceIndicators');
jest.mock('../../../thunks/fetchDevelopmentIndicators');


describe('AddComparisonForm', () => {

	let mockInput;
	let mockEvent
	let mockHistory;
  let mockLocation;
  let mockDataBase;
  let mockDataSet;
  let mockFetchDevelopmentIndicators;
  let mockFetchGovernanceIndicators;
  let mockSelectLocation;
  let additionalLocation;
  let wrapper;

  beforeEach(() => {

  	mockInput = { target: { value: 'Nigeria'} };

   	mockEvent = { preventDefault: jest.fn() };

  	mockHistory = [];

    mockLocation = [{
      name:'Burkina Faso',
      alpha_2:'BF',
      alpha_3:'BFA',
      country_code:'854',
      iso_3166_2:'ISO 3166-2: BF',
      region:'Africa',
      sub_region:'Sub-Saharan Africa',
      intermediate_region:'Western Africa',
      region_code:'002',sub_region_code:'202',
      intermediate_region_code:'011'
    }];

    additionalLocation =   {
	    name: 'Nigeria',
	    alpha_2: 'NG',
	    alpha_3: 'NGA',
	    country_code: '566',
	    iso_3166_2: 'ISO 3166-2: NG',
	    region: 'Africa',
	    sub_region: 'Sub-Saharan Africa',
	    intermediate_region: 'Western Africa',
	    region_code: '002',
	    sub_region_code: '202',
	    intermediate_region_code: '011'
	  };

    mockDataBase = 'WWDI';

    mockDataSet = 'SM_POP_REFG';

	  mockFetchDevelopmentIndicators = jest.fn();

	  mockFetchGovernanceIndicators = jest.fn();

	  mockSelectLocation = jest.fn();

    wrapper = shallow(
    	<AddComparisonForm 
    		dataBase={mockDataBase}
    		dataSet={mockDataSet}
    		location={mockLocation}
    		fetchGovernanceIndicators={mockFetchGovernanceIndicators}
    		fetchDevelopmentIndicators={mockFetchDevelopmentIndicators}
    		selectLocation={mockSelectLocation}
    		history={mockHistory}
    	/>
  	);

  });

	it('should match the snapshot', () => {

		expect(wrapper).toMatchSnapshot();

	});

  it('should have a default state of an empty object ', () => {

    const expected = {};

    const results = wrapper.state();

    expect(results).toEqual(expected);

  });

  it('should update the state when handleChange is invoked', () => {

    const expected = additionalLocation;

    wrapper.instance().handleChange(mockInput);

    const results = wrapper.state('additionalLocation');

    expect(results).toEqual(expected);

  });

   it('should invoke selectLocation when there is a additionalLocation property in local state', () => {

    wrapper.instance().handleChange(mockInput);

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockSelectLocation).toHaveBeenCalledWith([...mockLocation, additionalLocation]);

  });

   it('should not invoke selectLocation when there is a not additionalLocation property in local state', () => {

    mockInput = { target: { value: 'Naria'} };

   	wrapper.instance().handleChange(mockInput);

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockSelectLocation).not.toHaveBeenCalledWith([...mockLocation, additionalLocation]);

  });

   it('should invoke fetchGovernanceIndicators when the form is submitted with a database_code of WWDI', () => {

	  mockDataBase = 'WWGI'; 

    wrapper = shallow(
    	<AddComparisonForm 
    		dataBase={mockDataBase}
    		dataSet={mockDataSet}
    		location={mockLocation}
    		fetchGovernanceIndicators={mockFetchGovernanceIndicators}
    		fetchDevelopmentIndicators={mockFetchDevelopmentIndicators}
    		selectLocation={mockSelectLocation}
    		history={mockHistory}
    	/>
    );

   	wrapper.instance().handleChange(mockInput);

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockFetchGovernanceIndicators).toHaveBeenCalledWith([...mockLocation, additionalLocation], mockDataSet, mockDataBase);

  });

   it('should invoke fetchDevelopmentIndicators when the form is submitted with a database_code other than WWGI', () => {

   	wrapper.instance().handleChange(mockInput);

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockFetchDevelopmentIndicators).toHaveBeenCalledWith([...mockLocation, additionalLocation], mockDataSet, mockDataBase);

  });

   it('should direct the user to the stats page of the application', () => {

   	wrapper.instance().handleChange(mockInput);

    wrapper.instance().handleSubmit(mockEvent);

    expect(mockHistory).toEqual(['/stats']);

  });

	describe('mapStateToProps', () => {

		let mockState;
		let mockDataBase;
		let mockDataSet;

		beforeEach(() => {

			mockDataBase = {
    		name: 'Worldwide Governance Indicators',
    		database_code: 'WWGI'
    	}

    	mockDataSet = {
        name: 'Voice and Accountability', 
        dataset_code: 'VA'
      }

			mockState = {
				dataBase: mockDataBase,
				dataSet: mockDataSet,
				location: mockLocation
			};

		});

		it('should return an object with dataBase key', () => {

			const expected = {
				dataBase: mockDataBase.database_code,
				dataSet: mockDataSet.dataset_code,
				location: mockLocation
			};

			const mappedProps = mapStateToProps(mockState);

			expect(mappedProps).toEqual(expected);

		});

	});

	describe('mapDispatchToProps', () => {

		let actionToDispatch;
		let mockDispatch;

		beforeEach(() => {

			mockDispatch = jest.fn();

		});

		it('calls dispatch with the fetchDevelopmentIndicators action when called', () => {

	  	mockDataBase = 'WWGI'; 

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = fetchDevelopmentIndicators(mockLocation, mockDataBase, mockDataSet);

			mappedProps.fetchDevelopmentIndicators(mockLocation, mockDataBase, mockDataSet);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});

		it('calls dispatch with the fetchDevelopmentIndicators action when called', () => {

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = fetchGovernanceIndicators(mockLocation, mockDataBase, mockDataSet);

			mappedProps.fetchGovernanceIndicators(mockLocation, mockDataBase, mockDataSet);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});

		it('calls dispatch with the selectLocation action when called', () => {

			const mappedProps = mapDispatchToProps(mockDispatch);

			actionToDispatch = setLocation([...mockLocation, additionalLocation]);

			mappedProps.selectLocation([...mockLocation, additionalLocation]);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);

		});
		
	});

});