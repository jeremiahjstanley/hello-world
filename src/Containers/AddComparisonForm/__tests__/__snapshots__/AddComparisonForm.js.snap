// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddComparisonForm should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddComparisonForm
    dataBase="WWDI"
    dataSet="SM_POP_REFG"
    fetchDevelopmentIndicators={[Function]}
    fetchGovernanceIndicators={[Function]}
    history={Array []}
    location={
        Array [
            Object {
              "alpha_2": "BF",
              "alpha_3": "BFA",
              "country_code": "854",
              "intermediate_region": "Western Africa",
              "intermediate_region_code": "011",
              "iso_3166_2": "ISO 3166-2: BF",
              "name": "Burkina Faso",
              "region": "Africa",
              "region_code": "002",
              "sub_region": "Sub-Saharan Africa",
              "sub_region_code": "202",
            },
          ]
    }
    selectLocation={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          className="back-button"
          replace={false}
          to="/stats"
>
          ◀ 
          <span>
                    back
          </span>
</Link>,
        <form
          className="change-form"
          onSubmit={[Function]}
>
          <input
                    list="countries"
                    onChange={[Function]}
                    placeholder="enter a country"
                    type="text"
          />
          <datalist
                    id="countries"
          >
                    <option>
                              Afghanistan
                    </option>
                    <option>
                               Åland Islands
                    </option>
                    <option>
                              Albania
                    </option>
                    <option>
                              Algeria
                    </option>
          </datalist>
          <button
                    className="change-form-button"
                    disabled={true}
          >
                    Add
          </button>
</form>,
      ],
      "className": "change-form-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "◀ ",
            <span>
              back
</span>,
          ],
          "className": "back-button",
          "replace": false,
          "to": "/stats",
        },
        "ref": null,
        "rendered": Array [
          "◀ ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "back",
            },
            "ref": null,
            "rendered": "back",
            "type": "span",
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              list="countries"
              onChange={[Function]}
              placeholder="enter a country"
              type="text"
/>,
            <datalist
              id="countries"
>
              <option>
                            Afghanistan
              </option>
              <option>
                             Åland Islands
              </option>
              <option>
                            Albania
              </option>
              <option>
                            Algeria
              </option>
</datalist>,
            <button
              className="change-form-button"
              disabled={true}
>
              Add
</button>,
          ],
          "className": "change-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "list": "countries",
              "onChange": [Function],
              "placeholder": "enter a country",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option>
                  Afghanistan
</option>,
                <option>
                   Åland Islands
</option>,
                <option>
                  Albania
</option>,
                <option>
                  Algeria
</option>,
              ],
              "id": "countries",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "host",
                "props": Object {
                  "children": "Afghanistan",
                },
                "ref": null,
                "rendered": "Afghanistan",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "host",
                "props": Object {
                  "children": " Åland Islands",
                },
                "ref": null,
                "rendered": " Åland Islands",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "2",
                "nodeType": "host",
                "props": Object {
                  "children": "Albania",
                },
                "ref": null,
                "rendered": "Albania",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "3",
                "nodeType": "host",
                "props": Object {
                  "children": "Algeria",
                },
                "ref": null,
                "rendered": "Algeria",
                "type": "option",
              },
            ],
            "type": "datalist",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add",
              "className": "change-form-button",
              "disabled": true,
            },
            "ref": null,
            "rendered": "Add",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            className="back-button"
            replace={false}
            to="/stats"
>
            ◀ 
            <span>
                        back
            </span>
</Link>,
          <form
            className="change-form"
            onSubmit={[Function]}
>
            <input
                        list="countries"
                        onChange={[Function]}
                        placeholder="enter a country"
                        type="text"
            />
            <datalist
                        id="countries"
            >
                        <option>
                                    Afghanistan
                        </option>
                        <option>
                                     Åland Islands
                        </option>
                        <option>
                                    Albania
                        </option>
                        <option>
                                    Algeria
                        </option>
            </datalist>
            <button
                        className="change-form-button"
                        disabled={true}
            >
                        Add
            </button>
</form>,
        ],
        "className": "change-form-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "◀ ",
              <span>
                back
</span>,
            ],
            "className": "back-button",
            "replace": false,
            "to": "/stats",
          },
          "ref": null,
          "rendered": Array [
            "◀ ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "back",
              },
              "ref": null,
              "rendered": "back",
              "type": "span",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                list="countries"
                onChange={[Function]}
                placeholder="enter a country"
                type="text"
/>,
              <datalist
                id="countries"
>
                <option>
                                Afghanistan
                </option>
                <option>
                                 Åland Islands
                </option>
                <option>
                                Albania
                </option>
                <option>
                                Algeria
                </option>
</datalist>,
              <button
                className="change-form-button"
                disabled={true}
>
                Add
</button>,
            ],
            "className": "change-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "list": "countries",
                "onChange": [Function],
                "placeholder": "enter a country",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option>
                    Afghanistan
</option>,
                  <option>
                     Åland Islands
</option>,
                  <option>
                    Albania
</option>,
                  <option>
                    Algeria
</option>,
                ],
                "id": "countries",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "0",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Afghanistan",
                  },
                  "ref": null,
                  "rendered": "Afghanistan",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "host",
                  "props": Object {
                    "children": " Åland Islands",
                  },
                  "ref": null,
                  "rendered": " Åland Islands",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": "2",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Albania",
                  },
                  "ref": null,
                  "rendered": "Albania",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": "3",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Algeria",
                  },
                  "ref": null,
                  "rendered": "Algeria",
                  "type": "option",
                },
              ],
              "type": "datalist",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add",
                "className": "change-form-button",
                "disabled": true,
              },
              "ref": null,
              "rendered": "Add",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
